core/src/main/java/jenkins/util/UrlHelper.java

```diff
@@ -28,16 +28,42 @@
 import org.kohsuke.accmod.Restricted;
 import org.kohsuke.accmod.restrictions.NoExternalUse;
 
+/**
+ * Objective is to validate an URL in a lenient way sufficiently strict to avoid too weird URL
+ * but to still allow particular internal URL to be accepted
+ */
 @Restricted(NoExternalUse.class)
 public class UrlHelper {
     /**
-     * Authorize the {@code _} and {@code -} characters in domain
+     * Authorize the {@code _} and {@code -} characters in <strong>domain</strong>
+     * <p>
+     * Avoid {@code -} to be first or last, and {@code .} to be first (but can be last)
      * <p>
-     * Avoid {@code -} and {@code .} to be first or last
+     * 
+     * Lenient version of: <ol>
+     * <li> <a href="https://tools.ietf.org/html/rfc952">RFC-952</a> GRAMMATICAL HOST TABLE SPECIFICATION</li>
+     * <li> <a href="https://www.ietf.org/rfc/rfc1034.txt">RFC-1034</a> 3.5</li>
+     * <li> <a href="https://www.ietf.org/rfc/rfc1738.txt ">RFC-1738</a>3.1, host</li>
+     * <li> <a href="https://tools.ietf.org/html/rfc1123">RFC-1123</a> 2.1</li>
+     * </ol>
+     * <p>
+     * 
+     * Deliberately allow: <ol>
+     * <li> short domain name (often there are rules like minimum of 3 characters)</li>
+     * <li> long domain name (normally limit on whole domain of 255 and for each subdomain/label of 63)</li>
+     * <li> starting by numbers (disallowed by RFC-952 and RFC-1034, but nowadays it's supported by RFC-1123)</li>
+     * <li> use of underscore (not explicitly allowed in RFC but could occur in internal network, we do not speak about path here, just domain)</li>
+     * <li> custom TLD like "intern" that is not standard but could be registered locally in a network</li>
+     * </ol>
      */
     private static String DOMAIN_REGEX = System.getProperty(
             UrlHelper.class.getName() + ".DOMAIN_REGEX", 
-            "(\\w(\\.?-?\\w+)*)(:\\d{1,5})?"
+            "^" + 
```

Comment:
I'm liking the commented regexp here. If you use `Pattern.COMMENTS` as a flag, you can add inline comments with `#` and whitespace is ignored, though perhaps this feature doesn't work too well without string literals.

Reply:
That was my idea initially but in such case I needed to change the "simple" `.matches(` call into some `Pattern.compile` etc.
As it could be provided by System Property, I am not sure about the potential impact of the `COMMENTS` flag / usability.

Reply:
You can enable it inline by using `(?x)` to go into comment mode and `(?-x)` to leave it.